metasmoke.py

```diff
@@ -32,16 +32,98 @@
+                if Metasmoke.AutoSwitch.ping_failure_counter < 0:
+                    # Consecutive counter. Switch sign.
+                    Metasmoke.AutoSwitch.ping_failure_counter = 0
+                Metasmoke.AutoSwitch.ping_failure_counter += 1
+                current_counter = Metasmoke.AutoSwitch.counter
+                current_auto = Metasmoke.AutoSwitch.autoswitch_is_on
+            # MAX_FAILURES is constant so no lock.
+            if current_counter > Metasmoke.AutoSwitch.MAX_FAILURES and\
+               GlobalVars.MSStatus.is_up() and current_auto:
+                log("warning", "Last {} connections to metasmoke failed".format(current_counter) +
+                               " Setting metasmoke status to down.")
+                chatcommunicate.tell_rooms_with("debug", "**Warning**. {}:".format(GlobalVars.location) +
+                                                         " Last {} connection to metasmoke".format(current_counter) +
+                                                         " failed. Setting metasmoke status to **down**.")
+                Metasmoke.set_ms_down()
+
+        @staticmethod
+        def ping_succeeded():
+            """ Indicate a metasmoke status ping connection success """
+            with Metasmoke.AutoSwitch.rw_lock:
+                if Metasmoke.AutoSwitch.ping_failure_counter > 0:
+                    # Consecutive counter. Switch sign.
+                    Metasmoke.AutoSwitch.ping_failure_counter = 0
+                Metasmoke.AutoSwitch.ping_failure_counter -= 1
+                # Negative values for success
+                current_counter = -Metasmoke.AutoSwitch.ping_failure_counter
+                current_auto = Metasmoke.AutoSwitch.autoswitch_is_on
+            # MAX_SUCCESSES is constant so no lock.
+            if current_counter > Metasmoke.AutoSwitch.MAX_SUCCESSES and\
+               GlobalVars.MSStatus.is_down() and current_auto:
+                # Why use warning? Because some action may be needed if people don't think metasmoke is up.
+                log("warning", "Last {} connections to metasmoke succeeded".format(current_counter) +
+                               " Setting metasmoke status to up.")
+                chatcommunicate.tell_rooms_with("debug", "**Notice**. {}:".format(GlobalVars.location) +
+                                                         " Last {} connection to metasmoke".format(current_counter) +
+                                                         " succeeded. Setting metasmoke status to **up**.")
+                Metasmoke.set_ms_up()
+
+        @staticmethod
+        def enable_autoswitch(to_enable):
+            """ Enable or disable auto status switch """
+            with Metasmoke.AutoSwitch.rw_lock:
+                Metasmoke.AutoSwitch.autoswitch_is_on = to_enable
+            switch_auto_msg = "Metasmoke status autoswitch is now {}abled.".format("en" if to_enable else "dis")
+            log("info", switch_auto_msg)
```

Comment:
Let's only `log` and `tell` the debug rooms if we're actually changing state. We should probably only `log` and `tell` when the state actually changes (i.e. not log when we're changing it to the same state that it was already in) for `enable_autoswitch`, `set_ms_up` and `set_ms_down`

Reply:
Hopefully. Let's have the counter issue fixed first though :-P

Reply:
As otherwise the git tree will become very messy